File: app/utils/encryption.util.ts
import { scryptSync, randomBytes, createCipheriv, createDecipheriv } from 'crypto';

type EncryptedData = {
    iv: string;
    encryptedData: string;
};

const ALGORITHM: string = 'aes-256-cbc';
class EncryptionUtil {
    private static salt: string = 'salt';

    static encrypt(text: string, secretKey: string): string {
        const iv = randomBytes(16);
        const keyDerivation = scryptSync(secretKey, 'salt', 32);
        const cipher = createCipheriv(ALGORITHM, keyDerivation, iv);
        let encrypted = cipher.update(text, 'utf8', 'hex');
        encrypted += cipher.final('hex');
        return iv.toString('hex') + encrypted;
    }

    static decrypt(encrypted: string, secretKey: string): string {
        const iv = Buffer.from(encrypted.substring(0, 32), 'hex');
        const keyDerivation = scryptSync(secretKey, 'salt', 32);
        const encryptedText = encrypted.substring(32);
        const decipher = createDecipheriv(ALGORITHM, keyDerivation, iv);
        let decrypted = decipher.update(encryptedText, 'hex', 'utf8');
        decrypted += decipher.final('utf8');
        return decrypted;
    }
}

export default EncryptionUtil;

File: finally-bis-cron-gradle/src/main/java/com/nine15/fbis/utils/EncryptionUtil.java
package com.nine15.fbis.utils;

import javax.crypto.Cipher;
import javax.crypto.SecretKeyFactory;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.PBEKeySpec;
import javax.crypto.spec.SecretKeySpec;
import java.security.SecureRandom;
import java.security.spec.KeySpec;
import java.util.Base64;

public class EncryptionUtil {
    private static final String ALGORITHM = "AES/CBC/PKCS5Padding";
    private static final String KEY_ALGORITHM = "PBKDF2WithHmacSHA256";
    private static final String SALT = "salt";
    private static final int ITERATION_COUNT = 65536;
    private static final int KEY_LENGTH = 256;

    public static String encrypt(String text, String secretKey) throws Exception {
        SecureRandom random = new SecureRandom();
        byte[] iv = new byte[16];
        random.nextBytes(iv);
        IvParameterSpec ivParameterSpec = new IvParameterSpec(iv);

        SecretKeySpec secretKeySpec = generateSecretKey(secretKey, SALT);
        Cipher cipher = Cipher.getInstance(ALGORITHM);
        cipher.init(Cipher.ENCRYPT_MODE, secretKeySpec, ivParameterSpec);

        byte[] encrypted = cipher.doFinal(text.getBytes());
        return Base64.getEncoder().encodeToString(iv) + Base64.getEncoder().encodeToString(encrypted);
    }

    public static String decrypt(String encryptedData, String secretKey) throws Exception {
        byte[] ivAndCipherText = Base64.getDecoder().decode(encryptedData);
        byte[] iv = new byte[16];
        System.arraycopy(ivAndCipherText, 0, iv, 0, 16);
        IvParameterSpec ivParameterSpec = new IvParameterSpec(iv);

        byte[] cipherText = new byte[ivAndCipherText.length - 16];
        System.arraycopy(ivAndCipherText, 16, cipherText, 0, cipherText.length);

        SecretKeySpec secretKeySpec = generateSecretKey(secretKey, SALT);
        Cipher cipher = Cipher.getInstance(ALGORITHM);
        cipher.init(Cipher.DECRYPT_MODE, secretKeySpec, ivParameterSpec);

        byte[] decrypted = cipher.doFinal(cipherText);
        return new String(decrypted);
    }

    private static SecretKeySpec generateSecretKey(String password, String salt) throws Exception {
        KeySpec spec = new PBEKeySpec(password.toCharArray(), salt.getBytes(), ITERATION_COUNT, KEY_LENGTH);
        SecretKeyFactory factory = SecretKeyFactory.getInstance(KEY_ALGORITHM);
        byte[] secretKey = factory.generateSecret(spec).getEncoded();
        return new SecretKeySpec(secretKey, "AES");
    }

}
